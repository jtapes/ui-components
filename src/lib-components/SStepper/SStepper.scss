@import '../../theme/styles/variables';
@import '../../theme/styles/mixins';
@import '../../theme/styles/reset';
@import '../../theme/styles/fonts';
@import '../../theme/styles/global';

$stepper--animate: 0.75s !default;

.s-stepper {
  display: flex;
  align-items: center;
  justify-content: space-between;

  &-item {
    position: relative;
    display: flex;
    justify-content: center;

    &__inner {
      display: flex;
      flex-direction: column;
      align-items: center;
    }

    &__title {
      transition: color $stepper--animate;
    }

    &__indicator {
      border-radius: 50%;
      width: $size--16;
      height: $size--16;
      transition: background-color $stepper--animate;
    }

    &__divider {
      position: absolute;
      left: calc(50% + #{$size--8});
      bottom: $size--8 - 1;
      width: 100%;
      height: 1px;

      &--filled & {
        &--color {
          width: 100%;
        }
      }

      &--color {
        width: 0;
        height: 100%;
        transition: width $stepper--animate;
      }
    }

    &:last-child {
      .s-stepper-item__divider {
        display: none;
      }
    }
  }
}
