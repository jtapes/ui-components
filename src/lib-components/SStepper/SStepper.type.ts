export enum EStepStatuses {
  WAIT = 'wait',
  ACTIVE = 'active',
  FINISH = 'finish',
}

export type SStepItemValue = number | string

export type SStepType = {
  title: string
  status: EStepStatuses
  value: SStepItemValue
}

export type SStepItemType = {
  title: string
  value: SStepItemValue
}
