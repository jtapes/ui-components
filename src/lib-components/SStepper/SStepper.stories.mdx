import { config, withDesign } from 'storybook-addon-designs'
import { withTests } from '@storybook/addon-jest';
import {
  Meta,
  Story,
  ArgsTable,
  Description,
  Source,
  Preview,
} from '@storybook/addon-docs'
import results from '/.jest-test-results.json'
import SStepper from '@/lib-components/SStepper/SStepper.vue'
import SButton from '@/lib-components/SButton/SButton.vue'

<Meta
  title="DESIGN SYSTEM/Atom/Stepper"
  component={SStepper}
  decorators={[withDesign, withTests({results})]}
  argTypes={{
    steps: {
      control: 'array',
      defaultValue: ['Корзина', 'Договор', 'Получение', 'Оплата']
    },
    value: {
      control: 'text',
      defaultValue: 'Договор',
      description: 'Выбранный шаг',
    }
  }}
  parameters={{
    design: config({
      type: 'figma',
      url:
        'https://www.figma.com/file/gsZV6tcRkd7YirnRoHY9nN/MVP-3-Market-Checkout-%E2%80%A2-Login?node-id=254%3A2316',
    }),
    docs: {
      source: {
        type: 'code',
      },
    },
  }}
/>

# Stepper

<Description of={SStepper} />

export const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { SStepper, SButton },
  template:
    `<div>
      <s-stepper v-model="value" :steps="steps" />
      <div class="d-f mt-80 jc-c">
        <s-button @click="changeStep('Корзина')">Корзина</s-button>
        <s-button @click="changeStep('Договор')">Договор</s-button>
        <s-button @click="changeStep('Получение')">Получение</s-button>
        <s-button @click="changeStep('Оплата')">Оплата</s-button>
      </div>
     </div>`,
  methods: {
    changeStep(value) {
      this.value = value
    }
  },
})

<Preview withToolbar withSource="none">
  <Story name="SStepper">
    {Template.bind({})}
  </Story>
</Preview>

## Пример

<Source
  language="html"
  code={`<s-stepper v-model="currStep" :steps="['Корзина', 'Договор', 'Получение', 'Оплата']" />`}
/>

<ArgsTable of={SStepper} />
