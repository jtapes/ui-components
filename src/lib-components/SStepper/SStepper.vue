<template>
  <div class="s-stepper">
    <div
      v-for="(step, index) in stepsArray"
      :key="index"
      class="s-stepper-item"
      :style="itemStyle"
    >
      <div class="s-stepper-item__inner">
        <span
          class="s-stepper-item__title bold mb-16"
          :class="{
            'color-primary': step.status === EStepStatuses.FINISH,
            'color-font': step.status === EStepStatuses.ACTIVE,
            'color-font-secondary': step.status === EStepStatuses.WAIT,
          }"
        >
          <slot name="step-title" :step="step">
            {{ step.title }}
          </slot>
        </span>
        <div
          class="s-stepper-item__indicator"
          :class="{
            'bg-primary': step.status === EStepStatuses.FINISH,
            'bg-danger': step.status === EStepStatuses.ACTIVE,
            'bg-border': step.status === EStepStatuses.WAIT,
          }"
        />
      </div>
      <div
        class="s-stepper-item__divider bg-border"
        :class="{
          's-stepper-item__divider--filled':
            step.status === EStepStatuses.FINISH,
        }"
      >
        <div class="s-stepper-item__divider--color bg-primary" />
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop, Model } from 'vue-property-decorator'
import { EStepStatuses, SStepType, SStepItemType, SStepItemValue } from './SStepper.type'

/** Компонент используется для отображения шагов */
@Component
export default class SStepper extends Vue {
  /** Массив шагов */
  @Prop({
    type: Array,
    default: () => [],
  })
  steps!: SStepItemType[] | string[]

  @Model('change')
  value!: SStepItemValue

  get stepsArray(): SStepType[] {
    let active: SStepItemValue | undefined
    return this.steps.map((step: SStepItemType | string) => {
      const stepValue = typeof step === 'object' ? step.value : step
      const stepTitle = typeof step === 'object' ? step.title : step
      let status: EStepStatuses

      if (active === undefined) {
        if (this.value !== stepValue) {
          status = EStepStatuses.FINISH
        } else {
          status = EStepStatuses.ACTIVE
          active = stepValue
        }
      } else {
        status = EStepStatuses.WAIT
      }

      return {
        title: stepTitle,
        status,
        value: stepValue,
      }
    })
  }

  get itemStyle(): { [key: string]: string } {
    return {
      'flex-basis': `${100 / this.steps.length}%`,
    }
  }

  EStepStatuses = EStepStatuses
}
</script>
