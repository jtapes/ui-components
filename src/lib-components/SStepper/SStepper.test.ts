import { mount } from '@vue/test-utils'
import SStepper from '@/lib-components/SStepper/SStepper.vue'

describe('SStepper', () => {
  test('Is Vue snapshot', async () => {
    const wrapper = mount(SStepper, {})
    wrapper.setProps({
      steps: ['Корзина', 'Договор', 'Получение', 'Оплата'],
      value: 'Договор',
    })

    await wrapper.vm.$nextTick()
    const steppers = wrapper.findAll('.s-stepper-item')
    expect(steppers).toHaveLength(4)
    expect(wrapper.element).toMatchSnapshot()
  })

  test('Another steps format', () => {
    const wrapper = mount(SStepper, {
      propsData: {
        steps: [
          {
            title: 'Корзина',
            value: 1
          },
          {
            title: 'Договор',
            value: 2
          },
          {
            title: 'Получение',
            value: 3
          },
          {
            title: 'Оплата',
            value: 4
          },
        ]
      }
    })

    const steppers = wrapper.findAll('.s-stepper-item')
    expect(steppers).toHaveLength(4)
  })
})
