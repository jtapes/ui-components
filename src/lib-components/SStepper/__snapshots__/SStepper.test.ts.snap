// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SStepper Is Vue snapshot 1`] = `
<div
  class="s-stepper"
>
  <div
    class="s-stepper-item"
    style="flex-basis: 25%;"
  >
    <div
      class="s-stepper-item__inner"
    >
      <span
        class="s-stepper-item__title bold mb-16 color-primary"
      >
        
          Корзина
        
      </span>
       
      <div
        class="s-stepper-item__indicator bg-primary"
      />
    </div>
     
    <div
      class="s-stepper-item__divider bg-border s-stepper-item__divider--filled"
    >
      <div
        class="s-stepper-item__divider--color bg-primary"
      />
    </div>
  </div>
  <div
    class="s-stepper-item"
    style="flex-basis: 25%;"
  >
    <div
      class="s-stepper-item__inner"
    >
      <span
        class="s-stepper-item__title bold mb-16 color-font"
      >
        
          Договор
        
      </span>
       
      <div
        class="s-stepper-item__indicator bg-danger"
      />
    </div>
     
    <div
      class="s-stepper-item__divider bg-border"
    >
      <div
        class="s-stepper-item__divider--color bg-primary"
      />
    </div>
  </div>
  <div
    class="s-stepper-item"
    style="flex-basis: 25%;"
  >
    <div
      class="s-stepper-item__inner"
    >
      <span
        class="s-stepper-item__title bold mb-16 color-font-secondary"
      >
        
          Получение
        
      </span>
       
      <div
        class="s-stepper-item__indicator bg-border"
      />
    </div>
     
    <div
      class="s-stepper-item__divider bg-border"
    >
      <div
        class="s-stepper-item__divider--color bg-primary"
      />
    </div>
  </div>
  <div
    class="s-stepper-item"
    style="flex-basis: 25%;"
  >
    <div
      class="s-stepper-item__inner"
    >
      <span
        class="s-stepper-item__title bold mb-16 color-font-secondary"
      >
        
          Оплата
        
      </span>
       
      <div
        class="s-stepper-item__indicator bg-border"
      />
    </div>
     
    <div
      class="s-stepper-item__divider bg-border"
    >
      <div
        class="s-stepper-item__divider--color bg-primary"
      />
    </div>
  </div>
</div>
`;
